sim/battle-queue.ts

```diff
@@ -0,0 +1,367 @@
+	/** speed of pokemon using move (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing the move */
+	pokemon: Pokemon;
+	/** location of the target, relative to pokemon's side */
+	targetLoc: number;
+	/** original target pokemon, for target-tracking moves */
+	originalTarget: Pokemon;
+	/** a move to use (move action only) */
+	moveid: ID;
+	/** a move to use (move action only) */
+	move: Move;
+	/** true if megaing or ultra bursting */
+	mega: boolean | 'done';
+	/** if zmoving, the name of the zmove */
+	zmove?: string;
+	/** if dynamaxed, the name of the max move */
+	maxMove?: string;
+	/** effect that called the move (eg Instruct) if any */
+	sourceEffect?: Effect | null;
+}
+
+/** A switch action */
+export interface SwitchAction {
+	/** action type */
+	choice: 'switch' | 'instaswitch';
+	order: 3 | 103;
+	/** priority of the action (lower first) */
+	priority: number;
+	/** speed of pokemon switching (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing the switch */
+	pokemon: Pokemon;
+	/** pokemon to switch to */
+	target: Pokemon;
+	/** effect that called the switch (eg U */
+	sourceEffect: Effect | null;
+}
+
+/** A Team Preview choice action */
+export interface TeamAction {
+	/** action type */
+	choice: 'team';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** unused for this action type */
+	speed: 1;
+	/** the pokemon switching */
+	pokemon: Pokemon;
+	/** new index */
+	index: number;
+}
+
+/** A generic action not done by a pokemon */
+export interface FieldAction {
+	/** action type */
+	choice: 'start' | 'residual' | 'pass' | 'beforeTurn';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** unused for this action type */
+	speed: 1;
+	/** unused for this action type */
+	pokemon: null;
+}
+
+/** A generic action done by a single pokemon */
+export interface PokemonAction {
+	/** action type */
+	choice: 'megaEvo' | 'shift' | 'runPrimal' | 'runSwitch' | 'event' | 'runUnnerve' | 'runDynamax';
+	/** priority of the action (lower first) */
+	priority: number;
+	/** speed of pokemon doing action (higher first if priority tie) */
+	speed: number;
+	/** the pokemon doing action */
+	pokemon: Pokemon;
+}
+
+export type Action = MoveAction | SwitchAction | TeamAction | FieldAction | PokemonAction;
+
+/**
+ * An ActionChoice is like an Action and has the same structure, but it doesn't need to be fully filled out.
+ *
+ * Any Action or ChosenAction qualifies as an ActionChoice.
+ *
+ * The `[k: string]: any` part is required so TypeScript won't warn about unnecessary properties.
+ */
+export interface ActionChoice {
+	choice: string;
+	[k: string]: any;
+}
+
+/**
+ * Kind of like a priority queue, although not sorted mid-turn in Gen 1-7.
+ *
+ * Sort order is documented in `BattleQueue.comparePriority`.
+ */
+export class BattleQueue extends Array<Action> {
+	battle: Battle;
+	constructor(battle: Battle) {
+		super();
+		this.battle = battle;
```

Comment:
The cute thing here is that whenever you call `queue.slice()` or `this.splice()` the return value is a new Queue that is constructed with a single parameter being the number of elements that will be returned. This makes trying to add stuff to the constructor interesting to say the least...

Reply:
Yeah, it was definitely a mistake to extend Array rather than wrap it. Feel free to refactor it to a wrapper instead.

Reply:
Oh the other hand, I would probably override `slice` to return a normal array, and `splice` to return itself.

Reply:
I looked into that but then you've got the ugliness of either referring to `battle.queue.queue` all the time or forwarding all of the Array methods like this:
```typescript
// Delegated array methods
[Symbol.iterator]() {
	return this.queue[Symbol.iterator]();
}
get [0]() {
	return this.queue[0];
}
entries() {
	return this.queue.entries();
}
item(index: number) {
	return this.queue[index];
}
get length() {
	return this.queue.length;
}
push(...actions: Action[]) {
	return this.queue.push(...actions);
}
shift() {
	return this.queue.shift();
}
slice() {
	return this.queue.slice();
}
splice(start: Number, end: Number, ...actions: Action[]) {
	this.splice(start, end, ...actions);
}
unshift(action: Action) {
	return this.queue.unshift(action);
}
```